client/packages/programs/src/JsonForms/stripEmptyAdditions.ts

```diff
@@ -0,0 +1,109 @@
+    if (obj.length === 0) return true;
+    return obj.every(it => objectOrArrayIsEmpty(it));
+  }
+
+  // object
+  if (Object.keys(obj).length === 0) return true;
+  const allValuesEmpty = Object.values(obj).every(it => {
+    if (typeof it === 'object') return objectOrArrayIsEmpty(it);
+    return it === undefined;
+  });
+  if (allValuesEmpty) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Recursively removes all empty data which has been added to the newData compared to the old
+ * input object.
+ *
+ * For example, given:
+ *
+ * old: { some: "value" }
+ * newData: { some: "value", obj: { add1: undefined }, array: [], add2: undefined}
+ *
+ * stripEmptyAdditions(old, newData) will return { some: "value" }.
+ */
+export const stripEmptyAdditions = (
+  old: JsonData | undefined,
+  newData: JsonData | undefined
+): JsonData | undefined => {
+  if (newData === undefined) return undefined;
+  if (!_.isObject(newData)) return newData;
+
+  if (_.isObject(newData) && !_.isArray(newData)) {
+    const object: JsonData = {};
+    const oldObj = !old || !_.isObject(old) || _.isArray(old) ? {} : old;
+
+    const allKeys = new Set<string>();
+    Object.keys(oldObj).reduce((prev, cur) => prev.add(cur), allKeys);
+    Object.keys(newData).reduce((prev, cur) => prev.add(cur), allKeys);
+    for (const key of allKeys) {
+      const o = oldObj[key];
+      let n = newData[key];
+      if (_.isObject(n) && !_.isArray(n)) {
+        n = stripEmptyAdditions(o, n);
+        if (objectOrArrayIsEmpty(n)) {
+          if (_.isEqual(o, {}) || _.isEqual(o, [])) {
+            // keep existing empty object
+            object[key] = o;
+          }
+          // ignore the empty addition
+          continue;
+        }
+      }
+      if (n !== undefined) object[key] = n;
+    }
+    if (Object.keys(object).length === 0) return undefined;
+    return object;
+  }
+
+  return newData;
+};
+
+// https://stackoverflow.com/questions/57874879/how-to-treat-missing-undefined-properties-as-equivalent-in-lodashs-isequalwit
+const isEqualIgnoreUndefined = (
+  a: JsonData | undefined,
+  b: JsonData | undefined
+) => {
+  const comparisonFunc = (a: JsonData | undefined, b: JsonData | undefined) => {
+    if (_.isArray(a) || _.isArray(b)) return;
+    if (!_.isObject(a) || !_.isObject(b)) return;
+
+    if (!_.includes(a, undefined) && !_.includes(b, undefined)) return;
+
+    // Call recursively, after filtering all undefined properties
+    return _.isEqualWith(
+      _.omitBy(a, value => value === undefined),
+      _.omitBy(b, value => value === undefined),
+      comparisonFunc
+    );
+  };
+  return _.isEqualWith(a, b, comparisonFunc);
+};
+
+export const isEqualIgnoreUndefinedAndEmpty = (
+  old: JsonData | undefined,
+  newData: JsonData | undefined
+) => {
+  const stripped = stripEmptyAdditions(old, newData);
+  // ignore undefined (e.g. in array objects which haven't been stripped) when comparing
+  return isEqualIgnoreUndefined(old, stripped);
```

Comment:
I first thought about stripping all `undefined` in `stripEmptyAdditions` but then used the exisiting `isEqualIgnoreUndefined` to just ignore undefines in array objects like [{value: undefined}]. Doing this, might be a bit faster performance wise because I mutate less(?) and thought it keeps `stripEmptyAdditions` smaller.

However, now I wonder if it would be better to do it all in `stripEmptyAdditions` and then compart the result like `_.isEqual(old, stripped`

Reply:
Thanks I just tried it but could find a good solution to apply it recursivly.  However, I simiplified the code a bit so now I just iterate over the keys of the new objects, i.e. now `reduce` needed anymore...

Reply:
Actually, I run into some more issues, I added some more test cases and had to add it back in :-( Can certainly be simplified but happy that all tests are passing now...